server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,530 @@
+    checkZkConnection(); // if ZK not connected, block, do not just return a cached value.
+    propStoreWatcher.registerListener(propCacheKey, this);
+
+    var props = cache.get(propCacheKey);
+    if (props != null) {
+      return props;
+    }
+
+    if (propCacheKey.getIdType() == PropCacheKey.IdType.SYSTEM) {
+      return new ConfigTransformer(zrw, codec, propStoreWatcher).transform(propCacheKey);
+    }
+
+    throw new PropStoreException(
+        "Invalid request for " + propCacheKey + ", the property node does not exist", null);
+  }
+
+  /**
+   * Convenience method for utilities that may not have a PropStore read the encoded properties
+   * directly from ZooKeeper. This allows utilities access when there is a ZooKeeper, may there may
+   * not be a full instance running. All exception handling is left to the caller.
+   *
+   * @param propCacheKey
+   *          the prop cache key
+   * @param watcher
+   *          a prop store watcher that will receive / handle ZooKeeper events.
+   * @param zrw
+   *          a ZooReaderWriter
+   * @return the versioned properties or null if the node does not exist.
+   * @throws IOException
+   *           if the underlying data from the ZooKeeper node cannot be decoded.
+   * @throws KeeperException
+   *           if a ZooKeeper exception occurs
+   * @throws InterruptedException
+   *           if the ZooKeeper read was interrupted.
+   */
+  public static @Nullable VersionedProperties readFromZk(final PropCacheKey propCacheKey,
+      final PropStoreWatcher watcher, final ZooReaderWriter zrw)
+      throws IOException, KeeperException, InterruptedException {
+    if (zrw.exists(propCacheKey.getPath())) {
+      Stat stat = new Stat();
+      byte[] bytes = zrw.getData(propCacheKey.getPath(), watcher, stat);
+      return codec.fromBytes(stat.getVersion(), bytes);
+    }
+    return null;
+  }
+
+  /**
+   * Copies all mappings from the specified map and into the existing property values and stores
+   * them into the backend store. New keys are added and keys the may have existed in the current
+   * properties are overwritten.
+   * <p>
+   * If multiple threads attempt to update values concurrently, this method will automatically
+   * retry. If the threads are setting different keys / values the result will be the sum of the
+   * changes. If the concurrent threads are attempting to set a value(s) for the same key(s), the
+   * value(s) will be the set to the values provided by the last thread to complete. The order is
+   * indeterminate.
+   *
+   * @param propCacheKey
+   *          the prop cache id
+   * @param props
+   *          a map of property k,v pairs
+   * @throws PropStoreException
+   *           if the values cannot be written or if an underlying store exception occurs.
+   */
+  @Override
+  public void putAll(PropCacheKey propCacheKey, Map<String,String> props) {
+    mutateVersionedProps(propCacheKey, VersionedProperties::addOrUpdate, props);
```

Comment:
I think the following would be save and may avoid some unnecessary work.   I noticed the FATE code that creates a table may sometimes call this code with an empty map.

```suggestion
  public void putAll(PropCacheKey propCacheKey, Map<String,String> props) {
    if(props.isEmpty())
       return;
    mutateVersionedProps(propCacheKey, VersionedProperties::addOrUpdate, props);
```

Reply:
resolved in e1b3b1a751